xds: Allow LB policy to be configured

dnephin opened the pull request:
Closes #6525, replaces #7728

This PR removes the LB types that use a hash, removing the need for a hash policy. It allows configuring the LB policy to `random` or `least_request`.

dnephin commented on agent/xds/server.go:
```diff
@@ -170,7 +170,7 @@ func (s *Server) StreamAggregatedResources(stream ADSStream) error {
 	err := s.process(stream, reqCh)
 	if err != nil {
-		s.Logger.Debug("Error handling ADS stream", "error", err)
+		s.Logger.Error("Error handling ADS stream", "error", err)
```
While debugging a failing integration test I noticed these errors were set to debug. 

I suspect there is a reason for this, but it caught me by surprise. I missed the error the first few times.  Is Debug the right level here? Error seems more appropriate, otherwise these errors would likely be missed in production deployments.

mkeeler replied:
This seems like the right level to me.

dnephin commented:
I will move the test to the new case as suggested in https://github.com/hashicorp/consul/pull/7713#discussion_r422238936 , once that PR merges, and also add the docs

dnephin commented:
Rebased this PR with master and added some of the changes from #7728.  I think what is missing is adding the HashPolicy to the route, and a test case for that change.

When I worked on #7728 it requires some refactoring to be able to set the hash policy on route, and now that refactoring heavily conflicts, so I think that might need to be re-done.
